aptos-move/framework/src/natives/function_info.rs

```diff
@@ -0,0 +1,158 @@
+) -> SafeNativeResult<(ModuleId, Identifier)> {
+    match arguments.pop_back() {
+        Some(val) => match val.value_as::<StructRef>() {
+            Ok(v) => {
+                let module_address = v
+                    .borrow_field(0)
+                    .and_then(|v| v.value_as::<Reference>())
+                    .and_then(|v| v.read_ref())
+                    .and_then(|v| v.value_as::<AccountAddress>())
+                    .map_err(SafeNativeError::InvariantViolation)?;
+
+                let module_name = identifier_from_ref(
+                    v.borrow_field(1)
+                        .map_err(SafeNativeError::InvariantViolation)?,
+                )?;
+
+                let func_name = identifier_from_ref(
+                    v.borrow_field(2)
+                        .map_err(SafeNativeError::InvariantViolation)?,
+                )?;
+                Ok((ModuleId::new(module_address, module_name), func_name))
+            },
+            Err(e) => return Err(SafeNativeError::InvariantViolation(e)),
+        },
+        None => {
+            return Err(SafeNativeError::InvariantViolation(PartialVMError::new(
+                StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+            )))
+        },
+    }
+}
+
+/***************************************************************************************************
+ * native fun check_dispatch_type_compatibility_impl
+ *
+ *   Returns true if the function argument types of rhs is the same as (arguments type of lhs || &FunctionInfo)
+ *   gas cost: base_cost + unit_cost * type_size
+ *
+ **************************************************************************************************/
+fn native_check_dispatch_type_compatibility_impl(
+    context: &mut SafeNativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert!(arguments.len() == 2);
+
+    // TODO: Figure out the correct gas charging schema here.
+    //
+    // We need to load the modules from lhs and rhs, and cloning the bytes for module id and function name.
+    context.charge(FUNCTION_INFO_CHECK_DISPATCH_TYPE_COMPATIBILITY_IMPL_BASE)?;
+
+    let rhs = {
+        let (module, func) = extract_function_info(&mut arguments)?;
+        context
+            .load_function(&module, &func)
+            .map_err(|_| SafeNativeError::Abort { abort_code: 2 })?
+    };
+    let lhs = {
+        let (module, func) = extract_function_info(&mut arguments)?;
+        context
+            .load_function(&module, &func)
+            .map_err(|_| SafeNativeError::Abort { abort_code: 2 })?
+    };
+
+    if lhs.parameter_types.is_empty() {
+        return Err(SafeNativeError::Abort { abort_code: 2 });
+    }
+
+    Ok(smallvec![Value::bool(
+        rhs.type_parameters == lhs.type_parameters
+            && rhs.return_types == lhs.return_types
+            && lhs.parameter_types[0 .. lhs.parameter_types.len() - 1] == rhs.parameter_types
```

Comment:
I actually don't see we check if the rhs is public function, I only see there's a check during runtime that the function is public. but what can happen is someone registers a private function with correct signature first then upgrades the function signature and visibility to public. not sure how bad it is, seems `address` and `signer` has different value representation but it smells very dangerous

Reply:
Great point. I added the check to make sure rhs has to be public